### ISSUE TITLE
Building guide?

### ISSUE THREAD

**Issue by marysiamzawka** [VinsCool/RASTER-Music-Tracker#8]
Would be nice to be able to compile the git version manually if one wishes to test bleeding edge features

**Comment by VinsCool** [comment]
Yeah, a lot of documentation is really lacking, this is one of the many things I would sincerely need help from other people interested in maintaining the project.
Anyway, for building the most recent git commit, it is not too complicated thankfully.

Recommended setup at the writing of this post:

- Visual Studio 2022
- C++20 (Latest)
- MFCv143 (Latest)

To build RMT, open the solution in Visual Studio, make sure to set the Build Target to Release, then build the entire Solution.
Both 32bit and 64bit versions will work, but I would recommend sticking to 64bit if it is possible.

Just a side note, this reply was written from memory, on mobile.
If there are any mistakes, I will make sure to edit my post later.

**Comment by peterdell** [comment]
Thanks Vin. I tested this and can confirm the release build compiles and starts just fine. The 32/64 bit debug build stop during start though. Not that I would a debugging something now...
![image](https://user-images.githubusercontent.com/10855380/234714376-5c17665a-57c9-45f5-b405-f00e2c2055e0.png)
![image](https://user-images.githubusercontent.com/10855380/234714417-5113679d-04ae-4619-81d3-d483ccef0b1a.png)

**Comment by VinsCool** [comment]
Hi Peter, thank your for your report!

About Debug build, yeah I am aware it's broken, but I have not actually used it myself so I have no idea currently what it happening with this.

I have personally been debugging from the Release build itself without any problem up to this point, so this is not really something to worry about.

I believe the actual reasons for crashing is related to the way debug stuff is being used, which is something I have literally never tried to use so that's pretty much the only reason why it is currently broken.

**Comment by peterdell** [comment]
I see, I was not aware the release build also includes the debug information and can be debugged.

**Comment by VinsCool** [comment]
Oh yeah, I am aware of the missing files, sorry for the confusion.
They were moved to a different folder, if I am not mistaken, else I think I still have a copy somewhere.

**Comment by ivop** [comment]
When I did the 64-bit port, debug builds still worked, so you must have broken something afterwards :) I looked into the commit messages, but they are very vague. I suggest you make smaller commits (single issues) and better commit messages, so it'll be easy to find regressions later.

You can indeed debug release builds, but it works better with debug builds. Hence the name ;) It uses a lower compiler optimization level and includes all symbols in the generated binary. A release build won't do that.

**Comment by VinsCool** [comment]
Fair point, I try to actually commit more often for that reason, haha!

As for debugging, I was indeed able to get my way around it despite using the Release builds.
I found an interesting softlock last night by accident, and could track it down without too much troubles, thankfully.

Speaking of which, I have no idea what I broke to get the Debug build to fail, so I might as go and try to fix it now that we mention that again, lol

**Comment by VinsCool** [comment]
In other news, I found exactly what I broke, I have no idea how I even missed that, I suppose the lack of sleep did weird things to my thought process.
[https://github.com/VinsCool/RASTER-Music-Tracker/commit/bcfe72be169ca91bbdcc6d14deae6595693aeed2](url)

The Debug build should work again now, haha!

**Comment by ivop** [comment]
Thanks for fixing it!

As for debugging release builds, IIUC VS keeps debug information in a separate database for release builds, so you still have some sort of symbols and references. But as you can see at Peter' s screenshot, it is not very helpful when aggressively optimizing is on. Reordering instructions and all that jazz results in it pointing to a simple boolean compare, which is obviously not what crashed the application :)

**Comment by TheMontezuma** [comment]
> The Debug build should work again now, haha!

It still does not work out of the box.
You need to adjust C++ Language Standard (it is set to C++ 14 in Debug configuration per default).

![2023-05-14 14_14_04-Rmt - Microsoft Visual Studio](https://github.com/VinsCool/RASTER-Music-Tracker/assets/14121988/6c75d516-dc6a-4f33-be3e-fd0345ae840d)

**Comment by VinsCool** [comment]
Thanks for the heads up, sorry about the technical issues, I am still learning how to do all of this programming stuff :D
